lightning/src/routing/router.rs

```diff
@@ -71,6 +71,21 @@ pub struct Route {
 	pub paths: Vec<Vec<RouteHop>>,
 }
 
+/// A [`RouteHop`] responsible for a failed payment in context of the path containing it.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FailingRouteHop {
+	/// A path that included the failing [`RouteHop`].
+	pub path: Vec<RouteHop>,
+
+	/// The index of the failing [`RouteHop`] within the path.
```

Comment:
The semantics of this are really confusing. Most of that is just cause `RouteHop` confusingly refers to both a node and a channel, which is itself confusing, but its compounded here.

In general, I think we have three choices here
 * take the complicated approach and have logic here to indicate if we're blaming a node or a channel.
 * take the moderate and decide which of the inbound or outbound channels to blame based on what the responding node said or,
 * take the easy approach and always blame the outbound channel from the node that we received the failure from.

I think the easy approach is basically what you do here, and ok for a firstpass - we'll never take the same path, which is good, but we may disable a channel that we really shouldn't.

In any case, I think this needs to be much clearer, and its unclear to me why we need this full struct when we could just return a short channel id.

Reply:
> In any case, I think this needs to be much clearer, and its unclear to me why we need this full struct when we could just return a short channel id.

We'll need the path for retry (to compute the amount), but I tend to agree it may be better to keep the path separate. It would at least be symmetry with the forthcoming event for a successful payment path.

As far as the scoring API goes, I had planned on giving a reference to the path and to the failing hop. But the latter could be the short channel id instead. Users could always scan the path if they want to find the matching hop, I suppose.

Reply:
Yea, I guess could go either way, but it does feel cleaner imo to just give the SCID instead of a reference to a path id - you're just going to go update the SCID in the graph anyway, right?